vulnerabilities/insufficient-access-control.md

```diff
@@ -1,3 +1,159 @@
+
+    constructor() {
+        owner = msg.sender;
+        managers[msg.sender] = true; // Assign the owner as manager
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Only the owner can call this function");
+        _;
+    }
+
+    modifier onlyManager() {
+        require(managers[msg.sender], "Only managers can call this function");
+        _;
+    }
+
+    // Functions with proper access control
+    function transferOwnership(address newOwner) public onlyOwner {
+        owner = newOwner;
+    }
+
+    function addManager(address newManager) public onlyOwner {
+        managers[newManager] = true;
+    }
+
+    function distributeTokens(address[] recipients) public onlyManager {
+        // Distribution logic here
+    }
+}
+
+```
+The ability to add ``managers`` allows the owner to delegate specific responsibilities without giving away full control. The ``onlyOwner`` and ``onlyManager`` modifiers ensure that only the appropriate roles can call specific functions, reducing potential security risks.
+
+#### 4. Whitelisting
+Whitelisting is used to specify a list of addresses or entities that are allowed to perform certain actions or access specific resources in a smart contract. By doing this, only those addresses that have been pre-approved have access, thereby reducing the attack surface and potential vulnerabilities.
+
+Example Implementattion:
+```solidity
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract AccessControl is ERC721, Ownable {
+    uint256 public tokenIdCounter;
+    mapping(address => bool) public whitelistedMinters;
+
+    constructor() ERC721("Collectibles", "COLL") {}
+
+    function mintNFT(address to) public onlyWhitelistedMinter {
+        tokenIdCounter++;
+        _mint(to, tokenIdCounter);
+    }
+
+    function addToWhitelist(address minter) public onlyOwner {
+        whitelistedMinters[minter] = true;
+    }
+
+    function removeFromWhitelist(address minter) public onlyOwner {
+        whitelistedMinters[minter] = false;
+    }
+
+    modifier onlyWhitelistedMinter() {
+        require(whitelistedMinters[msg.sender], "Only whitelisted minters can call this function");
+        _;
+    }
+}
+```
+In the above contract, we’ve used ``whitelisting`` by maintaining a mapping of addresses that are authorized to mint NFTs. The ``mintNFT`` function now has the ``onlyWhitelistedMinter`` modifier, which makes sure that only addresses on the whitelist can call this function. Additionally, the contract ``owner`` can manage the whitelist using the ``addToWhitelist`` and ``removeFromWhitelist`` functions.
+
+### Sources
+- https://metaschool.so/articles/access-control-vulnerabilities-in-smart-contracts/
+- https://medium.com/rektify-ai/how-to-mitigate-access-control-vulnerability-6df74c82af98
+- https://github.com/Quillhash/Solidity-Attack-Vectors/blob/main/data/1.md
```

Comment:
@0xSandyy, Could you please convert links to hyper links

Reply:
Thanks for the remainder!